3단계 - 자동차 경주

jaeykweon opened the pull request:
안녕하세요 3단계 자동차 경주 리뷰 요청 드립니다

docs는 [ReadMe](https://github.com/jaeykweon/kotlin-racingcar/tree/step3/src/main/kotlin/car_race)에서 확인하실 수 있습니다.

주어진 조건이 구체적이지 않아서 전체적인 로직 설계는 최대한 간단하게 진행하였습니다

CarPosition과  GameRound에 대한 단위 테스트는 진행하였는데,
Car, Cars 그리고 Main함수에 대한 테스트를 어떻게 해야할 지 모르겠어서 이 부분에 대한 질문을 드리고자 합니다.

감사합니다

jaeykweon committed:
refactor: 메소드 명 변경 (enter -> calculate)

jaeykweon committed:
refactor: [2단계] 메소드 명 변경 (enter -> calculate)

jaeykweon committed:
docs: 3단계 ReadMe 작성

jaeykweon committed:
docs: [3단계] ReadMe 작성

jaeykweon committed:
feature: car 도메인 구현

jaeykweon committed:
feature: [3단계] car 도메인 구현

jaeykweon committed:
refactor: 랜덤 생성 값 로직 분리

jaeykweon committed:
refactor: [3단계] 랜덤 생성 값 로직 분리

jaeykweon committed:
test: [3단계] car position 단위 테스트

jaeykweon committed:
test: car position 단위 테스트

jaeykweon committed:
feature: [3단계] gameRound 구현

jaeykweon committed:
feature: gameRound 구현

jaeykweon committed:
test: [3단계] GameRound 단위 테스트 작성

jaeykweon committed:
test: GameRound 단위 테스트 작성

jaeykweon committed:
feature: [3단계] UI 구현

jaeykweon committed:
feature: UI 구현

jaeykweon committed:
feature: main 함수 구현

jaeykweon committed:
feature: [3단계] main 함수 구현

ohgillwhan commented on src/test/kotlin/car_race/logic/car/CarPositionTest.kt:
```diff
@@ -0,0 +1,49 @@
+ */
+class CarPositionTest : FunSpec({
+
+    mockkObject(MovingSystem)
```
`mockk`를 실무에서 사용하기는 하는데, 미션을 진행하면서 사용하지 않고 해결하는 방법은 어떨까요?
https://victorydntmd.tistory.com/292

jaeykweon replied:
mock 객체를 활용하지 않을경우 예를들어 Car를 테스트 함에 있어서 CarPosition객체도 만들어줘야하는데, 이게 단위 테스트가 맞는지 의문이 있었습니다. 순수히 Car 객체의 흐름만 테스트 하는게 단위테스트의 목적성에 부합하지않을까 하고요.
그러면 Car 클래스 단위 테스트에 CarPosition 객체를 만들어 테스트하는것이 상관없을까요??

ohgillwhan replied:
넵 해당 테스트를 실행하기 위해서 다른 객체의 도움이 필요하다면, 생성해서 테스트해도 문제 없습니다.
실제로 프로덕션 환경에서도 다른 객체와 협력하여 프로그램이 실행하니, `CarPosition`을 만들어서 테스트 시키는 방법이 더 자연스러워 보입니다.

jaeykweon replied:
랜덤값의 경우 모킹을 하지않는다면 테스트를 어떻게 해야할까요??

jaeykweon replied:
위에 주신 링크는 전략 패턴에 관한 내용인데
혹시 테스트 할때 가짜 전략 (예를들어 무조건 4이상을 반환하는 전략 또는 4미만을 반환하는 전략)을 만들어서 테스트 하면 되는것일까요?

ohgillwhan replied:
넵 맞습니다!

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/832/commits/73de392bccfff25713bc412061ec7527564244d3)
에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/GameRound.kt:
```diff
@@ -0,0 +1,28 @@
+    companion object {
+        private const val INVALID_TYPE_MESSAGE = "게임 횟수는 양의 정수여야 합니다."
+
+        fun from(input: String): GameRound {
+            validate(input)
+
+            return GameRound(input.toLong())
+        }
+
+        private fun validate(input: String) {
+            try {
+                input.toLong()
+            } catch (e: NumberFormatException) {
+                throw IllegalArgumentException(
+                    INVALID_TYPE_MESSAGE,
+                    e
+                )
+            }
+            check(input.toLong() > 0) { throw IllegalArgumentException(INVALID_TYPE_MESSAGE) }
+        }
+    }
```
`toLong()`은 성능에 큰 영향을 주지 않지만, 3번을 호출하는건 약간 아쉬움이 남네요!
코틀린에서는 `try`,  `if`의 결과값을 변수로 담아서 사용할 수 있어요.
```suggestion
        fun from(input: String): GameRound {
            return GameRound(toLong(input.toLong))
        }

        private fun validate(input: String): Long {
            val number = try {
                input.toLong()
            } catch (e: NumberFormatException) {
                throw IllegalArgumentException(
                    INVALID_TYPE_MESSAGE,
                    e
                )
            }
            require(number > 0) { INVALID_TYPE_MESSAGE }
        }
    }
```

jaeykweon replied:
코드 감사합니다 주신 내용대로 수정해보겠습니다

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/832/commits/73de392bccfff25713bc412061ec7527564244d3)
에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Car.kt:
```diff
@@ -0,0 +1,19 @@
+package car_race.logic.car
+
+class Car private constructor(
+    private val carPosition: CarPosition = CarPosition()
+) {
```
`private constructor`를 사용하신 이유는 외부에서 잘못된 값이 안들어오거나, `Position`을 무조건 0부터 시작하고 싶으셔서 막으신거 같아요. 저는 객체가 생성될 때의 상태만 문제가 없다면, 상관이 없다고 생각합니다. 

예를들어 지갑이 있는데, 지갑의 돈이 0원 미만으로 객체를 생성하려고 한다거나 이런 케이스만 잘 막으면 되는거죠.

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Car.kt:
```diff
@@ -0,0 +1,19 @@
+        return carPosition
+    }
+
+    companion object {
+        fun from(): Car {
+            return Car()
+        }
+    }
```
생성자를 `public`으로 열고, 제거하는건 어떨까요?

jaeykweon replied:
우선 생성자를 private로 막고 팩토리인 from을 통해서 만들게 한 이유는
무조건 from에서의 검증구조를 거치도록 하기 위함입니다
여기서는 Car를 만듦에 있어서 검증 구조를 거칠 일이 없기 때문에 나중에 필요하면 다시 작성하는것으로하고
지우도록 하겠습니다

ohgillwhan replied:
팩터리 메서드를 통해서 검증하는것 보단, 객체가 생성될 때 호출되는 `init {}` 블록 안에서 점검하는게 제일 좋습니다.

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarPosition.kt:
```diff
@@ -0,0 +1,19 @@
+
+import car_race.logic.system.MovingSystem
+
+data class CarPosition(
+    private var value: Int = 0
+) {
```
`CarPosition`의 `value`는 가변이네요. 가변으로 생성하면 예상치 못한 곳에서 변경이 일어나서, 문제가 발생할 수 있습니다.
그래서 저는 내부의 값을 누가 바꿀지 예상하기 어려워서 최대한 `Car`와 같이 각각 식별이 가능한 객체인 `Entity`같은 것을 제외한 `data class`같은 성격의 클래스는 불변으로 만들고 사용하는것을 목표로 하고 있어요. 
https://mangkyu.tistory.com/131

jaeykweon replied:
처음에는 불변으로 하려했는데, 불변일 경우 차의 상태가 변화할때마다 새로운 차 포지션 객체를 만들어서 넘겨야 할텐데 이 방식이 맞을까요?

ohgillwhan replied:
위에 드린 링크에서 4번에 잘 쓰여져 있는데요, 불변 객체를 사용하면 코드가 더 안정하게 실행이 될 수 있으니 더 깔끔하다고 생각을 합니다.

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarPosition.kt:
```diff
@@ -0,0 +1,19 @@
+data class CarPosition(
+    private var value: Int = 0
+) {
+    fun getValue() = value
```
💯

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarPosition.kt:
```diff
@@ -0,0 +1,19 @@
+) {
+    fun getValue() = value
+
+    fun nextRound() {
+        val randomNum = MovingSystem.rollDice()
+
+        value += if (randomNum >= DIVIDE_NUMBER) 1 else 0
+    }
```
자동차가 움직이는 조건을 자동차에서 판단해야지, `CarPosition`이 판단하는것은 조금 아쉬운것 같아요!

jaeykweon replied:
그 부분에서 고민이 많았습니다
현실세계에서야 자동차가 움직이는것이지만
객체세계에서는 `CarPosition`이 변화하는 것이므로 `움직이라`라는 메세지만 열어두고 그 책임은 `CarPosition`이 져야하는 것이 아닌가 싶었습니다

jaeykweon replied:
그리고 주어진 조건 상 무조건 움직이는 것이 아니고 경우에 따라 가만히 있기도하기때문에 `CarPosition`에게 다음 라운드로의 변화만 알려주고 그에 따른 객체 내부의 변화는 `CarPosition`에게 책임지게 하였습니다
혹시 이 논리가 틀린것일까요??

ohgillwhan replied:
이미 우리는 객체를 사용하고 있고, 객체에 맞는 역할과 책임을 주는것이 쉽게 이해 가능한 코드라고 생각을 합니다.
쉽게 이해 가능한 코드는 우리의 예상을 크게 벗어나지 않는 코드라고 생각합니다.

간단히 텍스트로 풀어보겠습니다.
1. 자동차의 위치가 움직일 수 있는지 판단하여 움직인다.
2. 자동차가 움직일 수 있는지 판단하여 움직인다.

1, 2번의 차이는 앞에 나오는 "자동차의 위치", "자동차" 차이일 뿐인데요. 혹시 어느 역할이 움직임을 판단하는 책임을 가진다고 생각하실까요? 저는 개인적으로 2번이 매끄러운 생각이라고 생각합니다.

다시 요구사항을 자연스럽게 이야기해보겠습니다.
1. 자동차의 위치가 움직일지 말지 여부를 판단해서, 자동차의 위치를 한칸을 움직인다.
2. 자동차가 움직일지 말지 여부를 판단해서, 자동차의 위치를 한칸을 움직인다.

어느게 가장 자연스럽게 읽히실까요? 저는 다시 2번이라고 생각합니다. 그러면 2번의 코드를 한번 그대로 작성해보겠습니다.

```kotlin
class Car {
  fun move() { // 자동차 이동해라
      if (canMove()) { // 움직일 수 있으면
           carPosition.nextPosition() // 위치를 한칸 움직여라
     }
  }
  
  private fun canMove() {
      return MovingSystem.rollDice() >= DIVIDE_NUMBER)
  }
  
  companion object {
     const val DIVIDE_NUMBER = 4
  }
}
```

ohgillwhan replied:
움직일지 말지 여부는 자동차가 판단하고, 한 칸 이동하는것은 위치에게 맡기는 코드입니다.
이렇게 작성하면 각자의 위치에서 역할과 책임을 지킬 수 있다고 생각합니다.

jaeykweon replied:
자세한 설명과 코드 감사합니다
그런데 주신 코드의 경우 테스트 코드 작성할 때 의문이 생기는데
만약 `4이상의 경우 한칸 움직인다`와 `4미만인 경우 움직이지 않는다`를 테스트 하고싶은 경우는 어떻게해야할까요?
`canMove`가 private fun이라 테스트를 어떻게 해야할지 모르겠습니다,,

ohgillwhan replied:
`MovingSystem`을 `Object`로 선언하는게 아닌, 전략패턴을 사용해서 처리하면 될것 같습니다 :D

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/832/commits/73de392bccfff25713bc412061ec7527564244d3)
에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Car.kt:
```diff
@@ -0,0 +1,19 @@
+class Car private constructor(
+    private val carPosition: CarPosition = CarPosition()
+) {
+    fun nextRound() {
+        carPosition.nextRound()
+    }
```
`자동차 다음 라운드` 보다는 `자동차 움직여`가 더 이해하기 쉬울거같아요!

ohgillwhan replied:
자동차가 움직일 때 조건 체크하는것은 `CarPosition`보다는 `Car`에서 하는게 더 좋을것 같아요.
자동차 스스로의 위치는 자동차가 판단해야지, 위치라는 것이 판단하면 이상한것 같아요.

jaeykweon replied:
기존 코드에서 움직이는 것을 `CarPosition`이 책임졌기때문에 nextRound라고 하였습니다
move일 움직이는 지를 상위에서 정해줘야하는 절차지향적인 코드가 되지않을까 싶었습니다

ohgillwhan replied:
https://github.com/next-step/kotlin-racingcar/pull/832#discussion_r1020858714 에 남겼습니다~

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/832/commits/73de392bccfff25713bc412061ec7527564244d3)
에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -0,0 +1,41 @@
+package car_race.logic.car
+
+class Cars private constructor(
+    private val cars: List<Car>
+) {
```
여기도 `private` 접근 제어자를 풀어볼까요?

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/832/commits/73de392bccfff25713bc412061ec7527564244d3)
에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/system/MovingSystem.kt:
```diff
@@ -0,0 +1,8 @@
+
+    private val DICE_RANGE = (0..9)
+
+    fun rollDice(): Int = DICE_RANGE.random()
```
💯

ohgillwhan commented on src/main/kotlin/car_race/view/InputView.kt:
```diff
@@ -0,0 +1,19 @@
+import car_race.logic.GameRound
+import car_race.logic.car.Cars
+
+object InputView {
```
`object` 💯

ohgillwhan commented on src/main/kotlin/car_race/view/ResultView.kt:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    private fun printPosition(carPosition: CarPosition) {
+        println("-".repeat(carPosition.getValue() + 1))
```
`repeat` 💯

ohgillwhan commented on src/main/kotlin/step2/Main.kt:
```diff
@@ -18,7 +18,7 @@ fun main() {
     val calculatorInput = CalculatorInput.from(userCleanedInput)
 
-    val result = calculator.enter(calculatorInput)
+    val result = calculator.calculate(calculatorInput)
```
리뷰반영 👍

ohgillwhan reviewed (changes requested):
재용님 안녕하세요.
리뷰 확인 부탁드립니다! 궁금하신건 DM 부탁드려요~

jaeykweon committed:
Merge branch 'step3' of https://github.com/jaeykweon/kotlin-racingcar into step3

jaeykweon committed:
refactor: [3단계] 코드리뷰 적용

- movingSystem을 전략패턴으로 변경
- 불필요한 private 생성자 제거
- 중복된 toLong() 제거
- 단위 테스트도 코드 변경 사항에 맞게 변경

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -0,0 +1,43 @@
+) {
+
+    fun nextRound(movingSystem: MovingSystem) {
+        cars.map { car ->
```
`map`은 반복 후 값을 변환할 떄 사용합니다. 단순히 반복할 때는 `forEach` 가 어떨까요?

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/24acc80bbeb18d2ae06f3267cc259a934e7a1f72)에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    fun getPositions(): List<CarPosition> {
+        return cars.map { car ->
```
이렇게 반환할 때는 `map` 💯

ohgillwhan commented on src/main/kotlin/car_race/logic/system/MovingSystem.kt:
```diff
@@ -0,0 +1,5 @@
+package car_race.logic.system
+
+interface MovingSystem {
```
다형성 💯

ohgillwhan commented on src/main/kotlin/car_race/Main.kt:
```diff
@@ -0,0 +1,17 @@
+    val gameRound = InputView.inputGameRound()
+    val cars = InputView.inputCarCount()
+
+    val movingSystem = RandomMovingSystem
```
변수로 선언하지 않아도 괜찮을것 같아요!

jaeykweon replied:
가독성 때문에 이렇게 하였는데 제거하였습니다
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/24acc80bbeb18d2ae06f3267cc259a934e7a1f72)에 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/Main.kt:
```diff
@@ -0,0 +1,17 @@
+    val cars = InputView.inputCarCount()
+
+    val movingSystem = RandomMovingSystem
+    for (i in 1..gameRound.value) {
```
`repeat`은 어떨까요?

jaeykweon replied:
gameRound.value가 long타입인데 repeat는 int타입까지만 허용합니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -0,0 +1,43 @@
+
+import car_race.logic.system.MovingSystem
+
+class Cars(
```
테스트코드 작성해주세요~

jaeykweon replied:
[test: [4단계] Cars 단위 테스트 추가](https://github.com/next-step/kotlin-racingcar/pull/885/commits/bb183179cfd55f11df69d428ed1f280bf6c66206)에 작성하였습니다

ohgillwhan commented on src/main/kotlin/car_race/Main.kt:
```diff
@@ -0,0 +1,17 @@
+    for (i in 1..gameRound.value) {
+        ResultView.printEachRound(cars)
+        cars.nextRound(movingSystem)
+        println()
```
`ResultView`에 넣어도 괜찮아 보여요.

jaeykweon replied:
[refactor: [3단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/24acc80bbeb18d2ae06f3267cc259a934e7a1f72)에 적용하였습니다

ohgillwhan reviewed (approved):
재용님 코드 잘 봤습니다!
몇가지 리뷰 남겼는데, 확인해주세요!

누락된 테스트코드는 다음 미션에 추가 부탁드립니다!
